src/endpoints/oauth.rs

```diff
@@ -0,0 +1,1202 @@
+use crate::{
+    endpoints::{
+        append_device_id, convert_result, join_ids,
+        pagination::{paginate, Paginator},
+        BaseClient,
+    },
+    http::Query,
+    macros::{build_json, build_map},
+    model::*,
+    ClientResult, OAuth, Token,
+};
+
+use std::time;
+
+use log::error;
+use maybe_async::maybe_async;
+use rspotify_model::idtypes::PlayContextIdType;
+use serde_json::{json, Map};
+use url::Url;
+
+/// This trait implements the methods available strictly to clients with user
+/// authorization, including some parts of the authentication flow that are
+/// shared, and the endpoints.
+///
+/// Note that the base trait [`BaseClient`](crate::endpoints::BaseClient) may
+/// have endpoints that conditionally require authorization like
+/// [`user_playlist`](crate::endpoints::BaseClient::user_playlist). This trait
+/// only separates endpoints that *always* need authorization from the base
+/// ones.
+#[maybe_async(?Send)]
+pub trait OAuthClient: BaseClient {
+    fn get_oauth(&self) -> &OAuth;
+
+    /// Obtains a user access token given a code, as part of the OAuth
+    /// authentication. The access token will be saved internally.
+    async fn request_token(&mut self, code: &str) -> ClientResult<()>;
+
+    /// Refreshes the current access token given a refresh token. The obtained
+    /// token will be saved internally.
+    async fn refresh_token(&mut self, refresh_token: &str) -> ClientResult<()>;
+
+    /// Tries to read the cache file's token, which may not exist.
+    async fn read_token_cache(&mut self) -> Option<Token> {
+        let tok = Token::from_cache(&self.get_config().cache_path)?;
+
+        if !self.get_oauth().scope.is_subset(&tok.scope) || tok.is_expired() {
+            // Invalid token, since it doesn't have at least the currently
+            // required scopes or it's expired.
+            None
+        } else {
+            Some(tok)
+        }
+    }
+
+    /// Parse the response code in the given response url. If the URL cannot be
+    /// parsed or the `code` parameter is not present, this will return `None`.
+    fn parse_response_code(&self, url: &str) -> Option<String> {
+        let url = Url::parse(url).ok()?;
+        let mut params = url.query_pairs();
+        let (_, url) = params.find(|(key, _)| key == "code")?;
+        Some(url.to_string())
+    }
+
+    /// Tries to open the authorization URL in the user's browser, and returns
+    /// the obtained code.
+    ///
+    /// Note: this method requires the `cli` feature.
+    #[cfg(feature = "cli")]
+    fn get_code_from_user(&self, url: &str) -> ClientResult<String> {
```

Comment:
Not sure it's suitable to put `get_code_from_user` into `OAuthClient` trait, since we could get code without `token` and `cred` parameters. Same as `parse_response_code`.

Reply:
But `get_code_from_user` doesn't really make sense if you're not using oauth, right? You need the code for the authorization flows, which is not needed in a base client.